### ISSUE TITLE
MIGRATED: WebView stopLoading no longer working in onNavigationStateChange on Android (#15679)

### ISSUE THREAD

**Issue by jamonholmgren** [react-native-community/react-native-webview#8]
Migrated from https://github.com/facebook/react-native/issues/15679.

@tss101 says:

---

Environment
Environment:
OS: macOS Sierra 10.12.6
Node: 6.11.2
Yarn: 0.15.1
npm: 3.10.10
Watchman: 4.9.0
Xcode: Xcode 8.2 Build version 8C38
Android Studio: 2.2 AI-145.3537739

Packages: (wanted => installed)
react: 16.2.0 => 16.2.0
react-native: 0.53.3 => 0.53.3 (reproduced in react-native: 0.54.2, 0.55.4)

Steps to Reproduce
Create a Webview in 0.54.0, point it to a url with an embedded youtube video (iframe) or link
Call stopLoading in onNavigationStateChange
Expected Behavior
Newly clicked url stops loading

Actual Behavior
Url keeps loading

Reproducible Demo
export var TestWebContentView = createReactClass( {displayName: 'TestWebContentView',
    handleNavigationChange: function(navState) {
         console.log(navState)
         this.webview.stopLoading();
        }
    },
    render: function() {
        return (
            <View style={rootViewStyle} >
                <WebView
                ref={(ref) => { this.webview = ref; }}
                scrollEnabled={false}
                source={{uri: 'http://www.simplehtmlguide.com/livedemo.php?e=youtube2' }}
                startInLoadingState={false}
                mixedContentMode={'always'}
                scalesPageToFit={false}
                onNavigationStateChange={this.handleNavigationChange}
                allowsInlineMediaPlayback={true}
                automaticallyAdjustContentInsets={false}
                />
            </View>
        )
    }
})
I'm quite stumped by this, since it was working a week or so ago on the same Android device, on react-native 0.44. However, I reverted the react-native version back to 0.44 and it's still not working. Might a Chrome update be responsible? iOS version appears to work.

---

_More info in original thread_

**Comment by maxammann** [comment]
Can confirm that this is still not working on react-native 0.57.0.

I tested this by clicking on an image.

**Comment by WestFR** [comment]
Same problem on react-native 0.57

**Comment by maxammann** [comment]
@WestFR maybe you can workaround this problem with: https://github.com/react-native-community/react-native-webview/pull/107

**Comment by nikilmethew** [comment]
Can confirm that this is still not working.
In some device WebView stopLoading() is working and in some of the device WebView stopLoading() is not working, Url keeps loading.

**Comment by steveccable** [comment]
@maxammann I was able to use the improvement from your linked PR to work around this same problem (after upgrading to a version that has your change, of course).
That said, it still isn't great that `stopLoading` does not, in fact, stop the page from loading.  There may also be use cases for when someone would want `onNavigationStateChange` rather than `shouldStartLoadWithRequest`, but I haven't really looked enough into what each does under the hood and where they are in the lifecycle to say.
Long story short, this should probably still be fixed, but your provided workaround of using the other method does accomplish the intended behavior.  Thanks for both implementing that and suggesting it here!

**Comment by reneses** [comment]
+1 for this bug, looking forward to the fix!

In the meanwhile, I was able to simulate the behavior of `stopLoading()` (at least for my use case, inside the `onLoadStart` handler) using `goBack` (yup, quite hacky):

```js
onLoadStart={() => {
  ...
  Platform.select({
    ios: () => this.webview.current.stopLoading(),
    android: () => this.webview.current.goBack()
  })();
  ...
});
```

**Comment by Titozzz** [comment]
Thanks for the workaround 👍 . However this is still a hack and should be fixed haha

**Comment by ShashankMalviya1994** [comment]
Same problem on react-native 0.57.4, Issue still persists.

**Comment by Salakar** [comment]
Hey all; would someone be able to provide an up to date reproduction for this, the one in the original issue is from React Native 0.44 😫 

Thanks

**Comment by Ehesp** [comment]
Closing as no reproduction. If someone can provide a reproduction on the latest RN version that'd be helpful 👍 Happy to reopen with a reproduction.

**Comment by Adrianotiger** [comment]
A better way is to check if the link should be loaded before loadStart is called.
Using this: `onShouldStartLoadWithRequest={event => {if(event.url==="XXXX") return false; return true;}}`

**Comment by peterjreynoldsii** [comment]
@adruanotiger or even `onShouldStartLoadWithRequest={event => event.url !== 'XXXXX'}` to keep it really clean. But yes this helped me.

**Comment by gavrichards** [comment]
I'm finding this is still an issue. I'm trying to handle stacking webview screens using StackActions.push() to push the same route on to the stack. I've got it working fine on iOS, but on Android stopLoading() doesn't appear to do anything. Tapping a link in page A opens up page B in a new view on the stack, but underneath page B has also opened in the first view too. So when you tap back, you see the same page again, and from there there's no way to get back to the original page again. There doesn't seem to be any way to prevent this.

**Comment by gavrichards** [comment]
I've created a repository that allows for this issue to be reproduced. 
https://github.com/gavrichards/react-native-webview-stack
The important code is in `src/WebViewWrapper.js`
However I'm also now finding it happening occasionally on iOS too in this example. In addition, I sometimes get the stacked view opening twice when tapping a link.

**Comment by ldco2016** [comment]
my webview is actually being rendered by `WebViewAutoHeight`, but the issue is the same I think. If I try this:

```
<WebViewAutoHeight
                ref={ref => (this.webview = ref)}
                key={s.Body.substr(10)}
                source={{
                  //prettier-ignore
                  html: `<body style="font-family: -apple-system, Roboto, sans-serif; background-color: ${lightTheme.grey2} !important; color: ${v2Colors.charcoalDarkest}; font-size: ${moderateScale(14, 0.2)}px;">${s.Body}</body>`
                }}
                onNavigationStateChange={event => {
                  if (event.url !== uri) {
                    this.webview.stopLoading();
                    Linking.openURL(event.url);
                  }
                }}
              />
```

I get `undefined is not an object (evaluating '_this2.webview.stopLoading')`

**Comment by gavrichards** [comment]
@ldco2016 I suspect that is because the WebViewAutoHeight component doesn't expose a ref prop, so this.webview never gets set.

**Comment by ldco2016** [comment]
@gavrichards , any idea for some type of workaround? Or a way to get a similar outcome as with `this.webview.stopLoading():`? The hyperlinks open in a separate browser, but they are also opening in the webview in the background.

**Comment by gavrichards** [comment]
@ldco2016 Sorry, I have no idea - as you can see above, I've commented reporting an issue of my own with stopLoading() not working for me either. I'm tempted to open a new issue to try and get the attention of the project maintainer.
